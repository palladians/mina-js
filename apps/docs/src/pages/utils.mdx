# Utils [Don't waste time on implementing these yourself.]

We've implemented some utilities that you might find useful. You can find them in the `@mina-js/utils` package.

## Units

Utils library provides functions for unit conversion.

### formatMina

Formats micro-Mina to human-readable Mina value.

```typescript twoslash
import { formatMina } from '@mina-js/utils'

const mina = formatMina(5_000_000_000n)
// -> "5"
```

### parseMina

Parses human-readable Mina value to micro-Mina.

```typescript twoslash
import { parseMina } from '@mina-js/utils'

const mina = parseMina('5')
// -> 5_000_000_000n
```

### formatUnit

```typescript twoslash
import { formatUnits } from '@mina-js/utils'

const formatted = formatUnits(4200000000000n, 10)
// -> "420"
```

### parseUnits

```typescript twoslash
import { parseUnits } from '@mina-js/utils'

const parsed = parseUnits("420", 10)
// -> 4200000000000n
```

## Web Workers

Proof related computations can be heavy and can block the main thread. To avoid this, you can use Web Workers to run these computations in a separate thread. We've prepared a JSON-RPC protocol to easily connect the dots.

```typescript twoslash
// @filename: worker.ts
import { createRpcHandler } from "@mina-js/utils";

const { messageHandler } = createRpcHandler({
  methods: {
    ping: async () => 'pong',
  }
})

self.onmessage = messageHandler

// @filename: index.ts
import { createRpc } from "@mina-js/utils";

const worker = new Worker(new URL('./worker.ts', import.meta.url))
const rpc = createRpc({ worker })
const response = await rpc.request({
  method: 'ping',
  params: [],
})
```
