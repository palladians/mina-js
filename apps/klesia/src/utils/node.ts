import { KlesiaNetwork } from "@mina-js/klesia-utils";
import { match } from "ts-pattern";
import { z } from "zod";

const MINA_NETWORK = KlesiaNetwork.parse(process.env.MINA_NETWORK ?? "devnet");
const NODE_API_DEVNET = z
	.string()
	.parse(
		process.env.NODE_API_DEVNET ??
			"https://api.minascan.io/node/devnet/v1/graphql",
	);
const NODE_API_MAINNET = z
	.string()
	.parse(
		process.env.NODE_API_MAINNET ??
			"https://api.minascan.io/node/mainnet/v1/graphql",
	);
const NODE_API_ZEKO_DEVNET = z
	.string()
	.parse(process.env.NODE_API_ZEKO_DEVNET ?? "https://devnet.zeko.io/graphql");

export const getNodeApiUrl = () => {
	return match(MINA_NETWORK)
		.with("devnet", () => NODE_API_DEVNET)
		.with("mainnet", () => NODE_API_MAINNET)
		.with("zeko_devnet", () => NODE_API_ZEKO_DEVNET)
		.exhaustive();
};
