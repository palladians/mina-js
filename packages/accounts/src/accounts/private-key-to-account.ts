import {
	NullifierSchema,
	SignedFieldsSchema,
	SignedMessageSchema,
	SignedTransactionSchema,
} from "@mina-js/utils";
import MinaSigner from "mina-signer";
import type { PrivateKeyAccount } from "../types";
import { toAccount } from "./to-account";

export type PrivateKeyToAccountParams = {
	privateKey: string;
	network?: "mainnet" | "testnet";
};

/**
 * @description Creates an Account from a private key.
 *
 * @returns A Private Key Account.
 */
export function privateKeyToAccount({
	privateKey,
	network = "mainnet",
}: PrivateKeyToAccountParams): PrivateKeyAccount {
	const client = new MinaSigner({ network });
	const publicKey = client.derivePublicKey(privateKey);
	const account = toAccount({
		publicKey,
		async signMessage({ message }) {
			return SignedMessageSchema.parse(client.signMessage(message, privateKey));
		},
		async signTransaction(signable) {
			if ("transaction" in signable) {
				return SignedTransactionSchema.parse(
					client.signTransaction(signable.transaction, privateKey),
				);
			}
			return SignedTransactionSchema.parse(
				client.signTransaction(signable.command as never, privateKey),
			);
		},
		async createNullifier({ message }) {
			return NullifierSchema.parse(
				client.createNullifier(
					message.map((el) => BigInt(el)),
					privateKey,
				),
			);
		},
		async signFields({ fields }) {
			return SignedFieldsSchema.parse(
				client.signFields(
					fields.map((el) => BigInt(el)),
					privateKey,
				),
			);
		},
	});

	return {
		...account,
		source: "privateKey",
	} as PrivateKeyAccount;
}
