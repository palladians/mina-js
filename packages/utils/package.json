{
	"name": "@mina-js/utils",
	"version": "0.0.1",
	"type": "module",
	"module": "dist/index.js",
	"types": "dist/index.d.ts",
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"default": "./dist/index.js"
		}
	},
	"files": ["dist"],
	"scripts": {
		"build": "tsc --noEmit && tsup",
		"test": "bun test",
		"cleanup": "rimraf dist .turbo"
	},
	"dependencies": {
		"mina-signer": "3.0.7",
		"serialize-error": "^11.0.3",
		"superjson": "2.2.1"
	},
	"devDependencies": {
		"zod": "3.23.8"
	},
	"peerDependencies": {
		"typescript": "^5.0.0"
	}
}
