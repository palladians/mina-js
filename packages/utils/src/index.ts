export * from "./types";
export * from "./validation";
export { formatMina } from "./format-mina";
export { formatUnits } from "./format-units";
export { parseMina } from "./parse-mina";
export { parseUnits } from "./parse-units";
export { createRpc, createRpcHandler } from "./worker-rpc";
export * as Test from "./test/constants";
