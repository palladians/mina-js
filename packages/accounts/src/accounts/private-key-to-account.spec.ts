import { expect, it } from "bun:test";
import { Test } from "@mina-js/utils";
import { privateKeyToAccount } from "./private-key-to-account";

it("matches default values", () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	expect(account).toMatchSnapshot();
});

it("signs a message", async () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	const message = "hello world";
	const signedMessage = await account.signMessage({ message });
	expect(signedMessage).toMatchSnapshot();
});

it("signs a transaction", async () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	const transaction = {
		nonce: "1",
		from: "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
		to: "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
		amount: "3000000000",
		fee: "100000000",
	};
	const signedTransaction = await account.signTransaction({ transaction });
	expect(signedTransaction).toMatchSnapshot();
});

it("signs a zkapp command", async () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	const command = {
		zkappCommand: {
			accountUpdates: [],
			memo: "E4YM2vTHhWEg66xpj52JErHUBU4pZ1yageL4TVDDpTTSsv8mK6YaH",
			feePayer: {
				body: {
					publicKey: "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
					fee: "100000000",
					validUntil: "100000",
					nonce: "1",
				},
				authorization: "",
			},
		},
		feePayer: {
			feePayer: "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
			fee: "100000000",
			nonce: "0",
			memo: "Test",
		},
	};
	const signedTransaction = await account.signTransaction({ command });
	expect(signedTransaction).toMatchSnapshot();
});

it("creates a nullifier", async () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	const message = [1n, 2n, 3n];
	const nullifier = await account.createNullifier({
		message: message.map((el) => el.toString()),
	});
	expect(typeof nullifier.private.c).toBe("string");
});

it("signs fields", async () => {
	const account = privateKeyToAccount({
		privateKey: Test.accounts[0].privateKey,
	});
	const fields = [1n, 2n, 3n];
	const signedFields = await account.signFields({
		fields: fields.map((el) => el.toString()),
	});
	expect(signedFields).toMatchSnapshot();
});
