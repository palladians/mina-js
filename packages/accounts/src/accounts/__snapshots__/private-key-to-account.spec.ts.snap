// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`matches default values 1`] = `
{
  "createNullifier": [Function: AsyncFunction],
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signFields": [Function: AsyncFunction],
  "signMessage": [Function: AsyncFunction],
  "signTransaction": [Function: AsyncFunction],
  "source": "privateKey",
  "type": "local",
}
`;

exports[`signs a message 1`] = `
{
  "data": "hello world",
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": {
    "field": "26035766330205137973322965816888613275878847328888669700853142209707796713851",
    "scalar": "19081530095096912440933327980723294881571166019208938317545493475572505570273",
  },
}
`;

exports[`signs a transaction 1`] = `
{
  "data": {
    "amount": "3000000000",
    "fee": "100000000",
    "from": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
    "memo": "",
    "nonce": "1",
    "to": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
    "validUntil": "4294967295",
  },
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": {
    "field": "14481941291402553573872495627690275042434233928114864565201875644822981729719",
    "scalar": "22662322359617803596615419576835284136875294231266919676620712008667696490687",
  },
}
`;

exports[`signs fields 1`] = `
{
  "data": [
    "1",
    "2",
    "3",
  ],
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": "7mXCUvhLhFvG9ptrdfNceCrpThkCUyg1ct2z8uwY7eQbKz7UNmhv33TbuDjTznaypJtXRiMJyQWDnf27TH1FSXG7uJHTKAd9",
}
`;

exports[`signs a zkapp command 1`] = `
{
  "data": {
    "feePayer": {
      "fee": "100000000",
      "feePayer": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
      "memo": "Test",
      "nonce": "0",
      "validUntil": "null",
    },
    "zkappCommand": {
      "accountUpdates": [],
      "feePayer": {
        "authorization": "7mXWqNfmqMTM5uSCS2xLfsRBLTjGZKTtpEakdsrdQz1EUgYXogSvKxxtfGbBkqQ2mZRMA3uPAM8riaCF56pkqpZBLr2kNBLa",
        "body": {
          "fee": "100000000",
          "nonce": "0",
          "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
          "validUntil": null,
        },
      },
      "memo": "E4YVT4x3A9rUhmjkjGn8ZYBLZn7zK4cfvnMtBYZFdWkg37n2s3nrP",
    },
  },
  "publicKey": "B62qmWKtvNQTtUqo1LxfEEDLyWMg59cp6U7c4uDC7aqgaCEijSc3Hx5",
  "signature": "7mXWqNfmqMTM5uSCS2xLfsRBLTjGZKTtpEakdsrdQz1EUgYXogSvKxxtfGbBkqQ2mZRMA3uPAM8riaCF56pkqpZBLr2kNBLa",
}
`;
